nexus/src/app/sagas/instance_delete.rs

```diff
@@ -0,0 +1,106 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use super::ActionRegistry;
+use super::NexusActionContext;
+use super::NexusSaga;
+use crate::app::sagas::declare_saga_actions;
+use crate::context::OpContext;
+use crate::{authn, authz};
+use serde::Deserialize;
+use serde::Serialize;
+use steno::ActionError;
+
+// instance delete saga: input parameters
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Params {
+    pub serialized_authn: authn::saga::Serialized,
+    pub authz_instance: authz::Instance,
+}
+
+// instance delete saga: actions
+
+declare_saga_actions! {
+    instance_delete;
+    INSTANCE_DELETE_RECORD -> "no_result1" {
+        + sid_delete_instance_record
+    }
+    DELETE_NETWORK_INTERFACES -> "no_result2" {
+        + sid_delete_network_interfaces
+    }
+    DEALLOCATE_EXTERNAL_IP -> "no_result3" {
+        + sid_deallocate_external_ip
+    }
+}
```

Comment:
Should there be any rollback cases represented here? What happens if a step fails?

Reply:
There *could* be; I'm not 100% sure there *should* be? If we're able to delete the instance record, I think we should be "best-effort" rolling forward deletion, rather than attempting to undo the deletion. 

For example, suppose we cannot re-allocate the deleted resource: either the resource itself was exhausted, user-data was deleted, or a sled cooperating in the allocation is no longer around. What should we do?

By having deletion be a saga, it's at least *durable* - even if Nexus crashes, it'll finish the work when it comes back online - but I'm not sure about the value of "making deletion undoable" vs "making sure it rolls forward, whatever that means".

TL;DR: *Maybe*? But there is a lot of complexity down the road of undoing deletions that I'm not really sure how to handle yet.